### PR TITLE
Load the configuration file by default

### PR DESCRIPTION
**Pull request by mattcaswell**
Previously we only loaded the config file by default for libssl. Now we do it for libcrypto too.

This is an alternative approach to that shown in #9372.

Doing this revealed some problems in evp_test. The "Availablein" keyword was supposed to suppress tests if the required provider was not available. Unfortunately a bug was causing the tests to fail if that provider *was* available. It also revealed (perhaps not unexpectedly) that evp_test does not currently work when only the FIPS provider is loaded despite us attempting to run it like that. We were running evp_test twice (one with a config file for default+legacy, and once with a config for fips only). However neither config files were being loaded so we were ending up just running it twice for no reason. Now that loading the config file works by default (as of this PR), we also temporarily disable loading the FIPS only config for evp_test until evp_test can be fixed.

Finally we also fix the krb5 external test. That test was using legacy algorithms, so before running it we provide a config file to load the default and legacy providers. I've set this PR to run with extended tests. However I expect extended tests will still fail until #9476 is also merged.

### CODE REVIEW

File: doc/internal/man3/ossl_provider_new.pod [hunk 1]
```diff
@@ -81,14 +83,18 @@ times as ossl_provider_activate() has.
...
+Typically I<noconfig> should be 0.
 The provider object it finds has its reference count incremented.
 
 ossl_provider_new() creates a new provider object named I<name> and
 stores it in the provider object store, unless there already is one
 there with the same name.
 If there already is one with the same name, it's returned with its
 reference count incremented.
+The config file will be automatically loaded unless I<noconfig> is set.
+Typically I<noconfig> should be 0.
```
Comment (levitte) [inline doc/internal/man3/ossl_provider_new.pod]: Both here and above, it would be good if there was a little be more saying when it should be 1 and when it should be 0. Who will remember in a year?

Comment (mattcaswell) [reply doc/internal/man3/ossl_provider_new.pod]: Done - fixup pushed.

**Comment by mattcaswell** [comment]
Realised that I forgot to update the OPENSSL_INIT_LOAD_CONFIG documentation and to add a CHANGES entry. Done now.

@levitte please could you reconfirm?

**Comment by richsalz** [comment]
Should there be a way to disable this, and only use the compiled-in defaults?

Setting the config file to ```/dev/null``` would do it, but ```CONF_get1_default_config_file()``` glues together two pathname pieces (```X509_get1_default_cert_area()``` -- which returns ```OPENSSLDIR``` passed down from Makefile -- and ```OPENSSL_CONF```) you can only do a relative path.  Should the get default config file not do the glue part if OPENSSL_CONF is an absolute path?

Do you want a PR for that or will you add it to this?

**Comment by mattcaswell** [comment]
> Should there be a way to disable this, and only use the compiled-in defaults?

You mean should there be a way to disable loading the config file by default? You can do this already (at run time) by calling `OPENSSL_init_crypto(OPENSSL_INIT_NO_LOAD_CONFIG, NULL);`

There isn't a compile time flag for this - but I'm not sure that is necessary. We've already been loading the config file by default for libssl since 1.1.1, and no one has requested such a feature yet.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Comment by richsalz** [comment]
What about a PR to make the config file be an absolute path?

It was less important before because FIPS mode and FIPS modules can now be auto-loaded.

**Comment by mattcaswell** [comment]
> Setting the config file to /dev/null would do it, but CONF_get1_default_config_file() glues together two pathname pieces (X509_get1_default_cert_area() -- which returns OPENSSLDIR passed down from Makefile -- and OPENSSL_CONF) you can only do a relative path. Should the get default config file not do the glue part if OPENSSL_CONF is an absolute path?

Errr....either you are confused or I am. It is perfectly possible to set the OPENSSL_CONF environment variable to an absolute path. `CONF_get1_default_config_file()` queries the OPENSSL_CONF environment variable and if it is set then returns it with no further processing. If that environment variable isn't set then it glues together OPENSSLDIR and the OPENSSL_CONF *define* (not to be confused with the environment variable). The OPENSSL_CONF define is just "openssl.cnf".

In other words if the OPENSSL_CONF environment variable is not set then it looks in OPENSSLDIR/openssl.cnf.

**Comment by richsalz** [comment]
I don't trust environment variables in all environments.  I want the compiled-in path to be absolute, not relative and not something that can be influenced by an environment variable.

**Comment by mattcaswell** [comment]
> I don't trust environment variables in all environments. I want the compiled-in path to be absolute, not relative and not something that can be influenced by an environment variable.

If OPENSSLDIR is an absolute path then the config file path will also be absolute - so I'm not sure I understand your concern here?

Not trusting environment variable to influence the decision is an entirely different issue that I don't currently have an opinion on.

**Comment by richsalz** [comment]
I can't set the config file to an absolute path, like /dev/null or /etc/openssl.conf.  I can only set it to something relative to the entire openssl data directory.

**Comment by mattcaswell** [comment]
> I can only set it to something relative to the entire openssl data directory.

No. There are three ways (that I can think of) to set the config file location:

1) Set the environment variable. This gives you the ability to set absolute or relative paths
2) Use the default location. This is always OPENSSLDIR/openssl.cnf. You don't get the opportunity to define the relative part of this. You can only define the absolute OPENSSLDIR part as a compilation option.
3) Load the config file explicitly passing the filename to use. In this case the default location is ignored and the filename you pass is used as is. It can be absolute or relative.

**Comment by levitte** [comment]
[ahem] actually, you can give a relative path to `--openssldir`.  It will be relative to PREFIX, which in turn *must* be absolute, but still...

**Comment by richsalz** [comment]
How can I build openssl so that the default config file is /dev/null or /etc/openssl.config? From what I see, this is not possible because of the "glue" that I mentioned above.

**Comment by levitte** [comment]
Debian's package is configured with this: `--openssldir=/etc/ssl`
That means that the fallback file name becomes `/etc/ssl/openssl.cnf`
Do the same or something similar.

**Comment by levitte** [comment]
We don't have the concept of using /dev/null as the config file. We can, if it's important, but I gotta say that I don't see that as terribly useful.

**Comment by richsalz** [comment]
/dev/null was just a sample config file.  /etc/openssl.cnf, with the rest of the *cert data* data being elsewhere is the key point that seems to be getting lost here.  The fact that my original comment https://github.com/openssl/openssl/pull/9492#issuecomment-516934233 points to "get cert area" should have been clear.

Cert area is public data.  config files are not the same.

**Comment by levitte** [comment]
Cert area is a subdirectory. What stops you from having more restrictive permissions on the config file? (I would argue, though, that the default config file should be public)

**Comment by richsalz** [comment]
Either you see configuration as separate from "bag of certs" or you don't, I guess.

Locking down a directory is hard: even if you lock down the file permissions it is hard to ensure the file can't be unlinked/created, without affecting the other *cert data* directories to make it easy to update them, for example.

I'm talking about a two-line change:
```
 char *CONF_get1_default_config_file(void)
{
    char *file, *sep = "", *basename = OPENSSL_CONF;
    int len;

    if ((file = ossl_safe_getenv("OPENSSL_CONF")) != NULL)
        return OPENSSL_strdup(file);

    if (*basename == '/')
       return OPENSSL_strdup(basename)
...
```

**Comment by petrovr** [comment]
Rich Salz wrote:
> How can I build openssl so that the default config file is /dev/null or /etc/openssl.config? From what I see, this is not possible because of the "glue" that I mentioned above.

What is issue?
What about package manager to not install file . It could install 
foo.sample pr foo.defaults. Note that this is generic approach for all 
projects.

Some projects allow configuration to be specified. In OpenSSL case the 
is one environment variable. Just set it to

/dev/null

.

Regards,
Roumen Petrov

**Comment by richsalz** [comment]
I want to set an absolute path -- one starting at root -- for example /etc/openssl.cnf  I do not want to use environment variables because I want a security audit to look at the source and be able to know exactly what is going on.

I don't know how else to explain this any better.  See the code fragment above.
